lab_1_keywords_tfidf/main.py

```diff
@@ -4,106 +4,144 @@
     """
     Composes a frequency dictionary from the token sequence
-
     Parameters:
     tokens (List[str]): Token sequence to count frequencies for
-
     Returns:
     Dict: {token: number of occurrences in the token sequence} dictionary
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if (isinstance(tokens, list) and tokens
+            and all(isinstance(token, str) for token in tokens)):
+        freq_dict = {}
+        for token in tokens:
+            if token not in freq_dict:
+                freq_dict[token] = 1
+            else:
+                freq_dict[token] += 1
+        return freq_dict
+    return None
 
 
 def get_top_n(frequencies: dict[str, Union[int, float]], top: int) -> Optional[list[str]]:
     """
     Extracts a certain number of most frequent tokens
-
     Parameters:
     frequencies (Dict): A dictionary with tokens and
     its corresponding frequency values
     top (int): Number of token to extract
-
     Returns:
     List[str]: Sequence of specified length
     consisting of tokens with the largest frequency
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if (dict_type_check(frequencies, str, (int, float)) and frequencies
+            and isinstance(top, int) and top is not (True or False) and top > 0):
+
+        sorted_words = [token for token, freq in sorted(frequencies.items(), key=lambda token: token[1], reverse=True)]
+        top_words = sorted_words[:top]
+        return top_words
+    return None
 
 
 def calculate_tf(frequencies: dict[str, int]) -> Optional[dict[str, float]]:
     """
     Calculates Term Frequency score for each word in a token sequence
     based on the raw frequency
-
     Parameters:
     frequencies (Dict): Raw number of occurrences for each of the tokens
-
     Returns:
     dict: A dictionary with tokens and corresponding term frequency score
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if dict_type_check(frequencies, str, int):
+        words_num = sum(frequencies.values())
+        tf_dict = {token: (freq / words_num) for token, freq in frequencies.items()}
+        return tf_dict
+    return None
 
 
 def calculate_tfidf(term_freq: dict[str, float], idf: dict[str, float]) -> Optional[dict[str, float]]:
     """
     Calculates TF-IDF score for each of the tokens
     based on its TF and IDF scores
-
     Parameters:
     term_freq (Dict): A dictionary with tokens and its corresponding TF values
     idf (Dict): A dictionary with tokens and its corresponding IDF values
-
     Returns:
     Dict: A dictionary with tokens and its corresponding TF-IDF values
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if (dict_type_check(term_freq, str, float) and term_freq
+            and dict_type_check(idf, str, float)):
+        tfidf_dict = {}
+        for token in term_freq:
```

Comment:
could you use dictionary comprehension instead of cycle?

Reply:
done